### DISCUSSION TITLE
How to export the generated waveform as an audio file?

### DISCUSSION THREAD

**Discussion by JossPabloPC** [sinshu/meltysynth#34] [Q&A]
Once I have my sequencer rendered, how can I tranform it to a Wav file ? I am using this example: 

// Create the synthesizer.
var sampleRate = 44100;
var synthesizer = new Synthesizer("TimGM6mb.sf2", sampleRate);

// Read the MIDI file.
var midiFile = new MidiFile("flourish.mid");
var sequencer = new MidiFileSequencer(synthesizer);
sequencer.Play(midiFile, false);

// The output buffer.
var left = new float[(int)(sampleRate * midiFile.Length.TotalSeconds)];
var right = new float[(int)(sampleRate * midiFile.Length.TotalSeconds)];

// Render the waveform.
sequencer.Render(left, right);

**Comment by sinshu** [comment]
This is an example code to export the waveform as a WAV file. Note that you need NAudio to compile this code. 

https://github.com/sinshu/meltysynth/blob/092cf25a21fac9b2fcbfc6314ae66266fef45500/Workspace/Examples.cs#L153

**Reply by JossPabloPC** [reply]
Thanks ! Ill give it a try right now

**Reply by JossPabloPC** [reply]
After applying the Render method, I noticed that the left and right arrays are populated with zeros. I'm unsure if this is due to an issue with my MIDI file or the soundfont I'm using. I attempted to create a chord using the example provided, but the problem persists. Can you please advise me on what might be causing this issue?


`    public static void CreateWavFromMidi()
    {
        
        var sampleRate = 44100;
        var synthesizer = new Synthesizer("C:/Users/pcjos/Documents/Unity/Sindri_MusicLab/Sindri_WFC/Assets/MusicLab/Editor/FragmentCreation/TimGM6mb.sf2", sampleRate);

        // Read the MIDI file.
        var midiFile = new MeltySynth.MidiFile("C:/Users/pcjos/Documents/Unity/Sindri_MusicLab/Sindri_WFC/Assets/MusicLab/Editor/FragmentCreation/A.mid");
        var sequencer = new MidiFileSequencer(synthesizer);
        sequencer.Play(midiFile, false);

        // The output buffer.
        var left = new float[(int)(sampleRate * midiFile.Length.TotalSeconds)];
        var right = new float[(int)(sampleRate * midiFile.Length.TotalSeconds)];

        // Render the waveform.
        sequencer.Render(left, right);

        for (int i = 0; i < left.Length; i++)
        {
            if (left[i] != 0)
            {
                Debug.Log(left[i]);
            }
        }

        WriteWaveFile(left, right, sampleRate, "C:/Users/pcjos/Documents/Unity/Sindri_MusicLab/Sindri_WFC/Assets/MusicLab/Editor/FragmentCreation/Res.wav");
        AssetDatabase.Refresh();

    }`

**Reply by sinshu** [reply]
Your code looks fine. Could you share the MIDI file?

**Reply by JossPabloPC** [reply]
Sure, I am using various midi files. I am also adding the soundFont used. Github would allow me to upload them so I am sharing a drive link: [https://drive.google.com/drive/folders/14YxIxVj5sxeVD6pdWeadzVobiiki8KbC?usp=sharing]

**Reply by sinshu** [reply]
The combination of TimGM6mb.sf2 and A.mid worked in my environment 🤔
I don't know much about Unity, so it may be difficult to give further advice...

**Reply by JossPabloPC** [reply]
How strage, maybe is the .NET version Unity uses. Thank you so much for your help.

**Reply by CodeSmile-0000011110110111** [reply]
Note: when you create or modify a file, you don't refresh the entire AssetDatabase, scanning all (!) files for changes and unloading unused resources along the way. 

You only need to call AssetDatabase.ImportAsset(fileThatWasModified)

**Comment by nickgal** [comment]
If you are using Unity, there is a [SavWav gist (one of many forks)](https://gist.github.com/Vanlalhriata/dc5b703cca6f2e885557e1c2f54c376a) that can be used to add the wav header from an AudioClip and avoid adding all of NAudio.

```cs
sequencer.RenderInterleaved(buffer);

var clip = AudioClip.Create(trackName, lenSamples, channels, sampleRate, false);
clip.SetData(buffer, 0);

SavWav.Save(wavPath, clip);
```
